### PR TITLE
Check service status maintenance

### PR DESCRIPTION
**Pull request by alesh-dulal**
* Added check for status under maintenance
* checked for status code 503 or retry-after > 0 for the under maintenance status

### CODE REVIEW

File: monitor/src/common/messages.js [hunk 1]
```diff
@@ -11,6 +11,11 @@ const messages = {
   [STATUS_DOWN]: {
     title: 'Service is Down',
     text: (name) => `${name} went down.`
+  },
+
+  [STATUS_MAINTENANCE]: {
+    title: 'Service is under maintenance',
+    text: (name) => `${name} is under maintenance right now.`
```
Comment (aviskarkc10) [inline monitor/src/common/messages.js]: do we need the right now at the end?

File: monitor/src/services/status.js [hunk 2]
```diff
@@ -17,29 +20,42 @@ export const FALLBACK_HTTP_METHOD = http.HEAD;
...
     // try sending another request with a fallback method.
-    // TODO: Make fallback http method configurable using chill.yml
-    if (shouldRetry(err, method)) {
+    // return checkHostStatus(service, FALLBACK_HTTP_METHOD);
+
+    if (shouldRetry(statusCode, method)) {
       logger().debug(
-        `Got ${err.response.statusCode} error for ${method} request on service ${name}. ` +
-        `Now trying with the fallback method ${FALLBACK_HTTP_METHOD}`
+        `Got ${statusCode} error for ${method} request on service ${name}. ` +
```
Comment (aviskarkc10) [inline monitor/src/services/status.js]: This should be 
```
`Got ${statusCode} error for ${method} request on service ${name}. Now trying with the fallback method ${FALLBACK_HTTP_METHOD}.`
```

File: monitor/src/services/status.js [hunk 3]
```diff
@@ -58,22 +74,47 @@ export function getCheckInterval(status, min, max) {
...
-    (err.response.statusCode === HttpStatus.METHOD_NOT_ALLOWED ||
-     err.response.statusCode === HttpStatus.NOT_IMPLEMENTED) &&
+    (statusCode === HttpStatus.METHOD_NOT_ALLOWED ||
+      statusCode === HttpStatus.NOT_IMPLEMENTED) &&
     method !== FALLBACK_HTTP_METHOD
   );
 }
 
+/**
+ * Check if it should retry sending HTTP request.
```
Comment (aviskarkc10) [inline monitor/src/services/status.js]: this function is checking if the status code is 200 and not what the doc block says

File: monitor/src/services/status.js [hunk 4]
```diff
@@ -58,22 +74,47 @@ export function getCheckInterval(status, min, max) {
...
+ *
+ * @param   {string} statusCode
```
Comment (aviskarkc10) [inline monitor/src/services/status.js]: `statusCode` is not `string` it is `Number`

File: monitor/src/seeds/statuses.js [hunk 5]
```diff
@@ -10,6 +10,7 @@ export function seed(knex) {
   return knex('statuses').insert([
     { id: UNKNOWN, name: 'Unknown', description: 'Status of the service is Unknown.' },
     { id: UP, name: 'Up', description: 'Status of the service is Up.' },
-    { id: DOWN, name: 'Down', description: 'Status of the service is Down.' }
+    { id: DOWN, name: 'Down', description: 'Status of the service is Down.' },
+    { id: MAINTENANCE, name: 'Maintenance', description: 'Status of the service is Maintenance.' }
```
Comment (aviskarkc10) [inline monitor/src/seeds/statuses.js]: should this be `Status of the service is Under Maintenance.` @kabirbaidhya @mesaugat

Comment (kabirbaidhya) [reply monitor/src/seeds/statuses.js]: Status of the service is under maintenance.

File: monitor/src/common/messages.js (hunk 1 above)
Comment (kabirbaidhya) [reply monitor/src/common/messages.js]: No.

File: monitor/src/seeds/statuses.js (hunk 5 above)
Comment (aviskarkc10) [reply monitor/src/seeds/statuses.js]: While you are at it please alos change the description of the `UP` and `Down` status to:
`Status of the service is down.` and  `Status of the service is up.`

File: monitor/src/services/status.js [hunk 6]
```diff
@@ -17,29 +20,41 @@ export const FALLBACK_HTTP_METHOD = http.HEAD;
...
+      simple: false
+    });
 
-    return STATUS_UP;
-  } catch (err) {
     // If the original HTTP method was not allowed (405 Method Not Allowed)
     // try sending another request with a fallback method.
-    // TODO: Make fallback http method configurable using chill.yml
-    if (shouldRetry(err, method)) {
+    // return checkHostStatus(service, FALLBACK_HTTP_METHOD);
```
Comment (aviskarkc10) [inline monitor/src/services/status.js]: this commented should go inside the if block below

```js
if (shouldRetry(statusCode, method)) {
      logger().debug(
        `Got ${statusCode} error for ${method} request on service ${name}. Now trying with the fallback method ${FALLBACK_HTTP_METHOD}`
      );
     return checkHostStatus(service, FALLBACK_HTTP_METHOD);
    }

File: monitor/src/services/status.js [hunk 7]
```diff
@@ -17,29 +20,41 @@ export const FALLBACK_HTTP_METHOD = http.HEAD;
...
+
+    if (shouldRetry(statusCode, method)) {
+      logger().debug(
+        `Got ${statusCode} error for ${method} request on service ${name}. Now trying with the fallback method ${FALLBACK_HTTP_METHOD}`
+      );
+    }
+
+    if (checkUnderMaintenance(statusCode, headers.maintain, checkMaintain)) {
```
Comment (aviskarkc10) [inline monitor/src/services/status.js]: this should be `else if` and should be part of the above `if...else` thread

File: monitor/src/services/status.js [hunk 8]
```diff
@@ -58,22 +73,47 @@ export function getCheckInterval(status, min, max) {
   return status === STATUS_UP ? max : min;
 }
 
+/**
+ * Check if the system is under Maintenance.
+ * Check status code in response for 503.
+ * Check if in header if Maintain is 1.
```
Comment (aviskarkc10) [inline monitor/src/services/status.js]: this line is not clear. Also this function has no idea what header is. Maybe we should write something along the lines of
```
Check if the system is under Maintenance.
Return true if value of statusCode is 503 or maintain is '1' else returns false.
```

**Comment by kabirbaidhya** [comment]
@alesh-dulal Can you update the PR description to explain briefly what's done?

File: monitor/src/services/status.js [hunk 9]
```diff
@@ -58,22 +73,47 @@ export function getCheckInterval(status, min, max) {
...
+ *
+ * @param {Number} statusCode
+ * @param {Number} maintain
+ * @param {Boolean} checkMaintain
+ * @returns {Boolean}
+ */
+function checkUnderMaintenance(statusCode, maintain, checkMaintain = false) {
+  return (
+    statusCode === 503 || (checkMaintain && maintain === MAINTENANCE_VALUE)
```
Comment (kabirbaidhya) [inline monitor/src/services/status.js]: This should be the condition I guess:
```ts
return (statusCode === 503 && headers['Retry-After'] > 0);
```

File: monitor/src/services/status.js (hunk 7 above)
Comment (kabirbaidhya) [inline monitor/src/services/status.js]: I think this should be in the catch block, because the `try` block would throw error on 503 status code.

Comment (aviskarkc10) [reply monitor/src/services/status.js]: @alesh-dulal has used the param `simple: false` so errors will be handled in the try block.
https://github.com/request/request-promise#get-a-rejection-only-if-the-request-failed-for-technical-reasons

Comment (aviskarkc10) [reply monitor/src/services/status.js]: The reason we are doing this is, there might be a case where the server returns `statusCode: 200` but `maintain: 1` in the header.

Comment (kabirbaidhya) [reply monitor/src/services/status.js]: Let's just stick to general convention. Let's expect the services and applications to send 2 things when made "intentionally unavailable", status = 503 AND Retry-After = X (X > 0)

**Comment by kabirbaidhya** [comment]
@alesh-dulal @aviskarkc10 @mesaugat Let's just stick to general convention; let's expect the services and applications to send 2 things when made "intentionally unavailable", status = 503 AND Retry-After = X (X > 0)
cc @leapfrogtechnology/chill-team

**Comment by kabirbaidhya** [comment]
This is how it should function:
```
IF (status = 503 AND Retry-After > 0) THEN 
  result = UNDER_MAINTENANCE

ELSE IF  (status = 200) THEN
  result = UP

ELSE IF (retried_times > MAX_RETRIES AND status = NON-200) THEN
  result = DOWN

ELSE 
  result = DOWN
```

File: monitor/src/services/status.js [hunk 10]
```diff
@@ -58,18 +64,29 @@ export function getCheckInterval(status, min, max) {
   return status === STATUS_UP ? max : min;
 }
 
+/**
+ * Check if the system is under maintenance.
+ * Return true if value of statusCode is 503 or check retryAfter is greater than 0 else returns false.
```
Comment (alesh-dulal) [inline monitor/src/services/status.js]: `Return true if value of statusCode is 503 or retryAfter is greater than 0 else return false.`

File: monitor/src/services/status.js [hunk 11]
```diff
@@ -28,18 +29,23 @@ export async function checkHostStatus(service, method = http.OPTIONS) {
...
+    if (shouldRetry(statusCode, method)) {
       logger().debug(
-        `Got ${err.response.statusCode} error for ${method} request on service ${name}. ` +
-        `Now trying with the fallback method ${FALLBACK_HTTP_METHOD}`
+        `Got ${statusCode} error for ${method} request on service ${name}. Now trying with the fallback method ${FALLBACK_HTTP_METHOD}`
       );
 
       return checkHostStatus(service, FALLBACK_HTTP_METHOD);
-    }
+    } else if (checkUnderMaintenance(statusCode, err.response.headers['retry-after'])) {
```
Comment (alesh-dulal) [inline monitor/src/services/status.js]: convert `err.response.headers['retry-after']` to number first

File: monitor/src/common/messages.js [hunk 12]
```diff
@@ -11,6 +11,11 @@ const messages = {
   [STATUS_DOWN]: {
     title: 'Service is Down',
```
Comment (kabirbaidhya) [inline monitor/src/common/messages.js]: down

File: monitor/src/monitoring/Monitor.js [hunk 13]
```diff
@@ -64,8 +64,8 @@ class Monitor {
    * Start monitoring services.
    */
   async startMonitoring() {
-    const { url, name, maxRetry, minInterval, maxInterval } = this.config;
-    const status = await statusService.checkHostStatus({ url, name });
+    const { url, name, maxRetry, minInterval, maxInterval, checkMaintain } = this.config;
```
Comment (kabirbaidhya) [inline monitor/src/monitoring/Monitor.js]: Let's not add an additional config `checkMaintain` - we'll make it opinionated.

File: monitor/src/monitoring/Monitor.js [hunk 14]
```diff
@@ -64,8 +64,8 @@ class Monitor {
...
+    const status = await statusService.checkHostStatus({ url, name, checkMaintain });
```
Comment (aviskarkc10) [inline monitor/src/monitoring/Monitor.js]: here as well. Remove the `checkMaintain`

File: monitor/src/services/status.js [hunk 15]
```diff
@@ -58,18 +65,29 @@ export function getCheckInterval(status, min, max) {
   return status === STATUS_UP ? max : min;
 }
 
+/**
+ * Check if the system is under maintenance.
+ * Return true if value of statusCode is 503 or retryAfter is greater than 0 else return false.
```
Comment (aviskarkc10) [inline monitor/src/services/status.js]: add `and` here as well

File: monitor/src/services/status.js [hunk 16]
```diff
@@ -58,18 +65,29 @@ export function getCheckInterval(status, min, max) {
...
  * Check if it should retry sending HTTP request.
  *
- * @param   {Error} err
- * @param   {string} method
+ * @param   {Number} statusCode
+ * @param   {String} method
  * @returns {Boolean}
  */
-function shouldRetry(err, method) {
+function shouldRetry(statusCode, method) {
```
Comment (kabirbaidhya) [inline monitor/src/services/status.js]: Good :+1:

File: monitor/src/services/status.js [hunk 17]
```diff
@@ -28,18 +29,24 @@ export async function checkHostStatus(service, method = http.OPTIONS) {
 
     return STATUS_UP;
   } catch (err) {
+    const statusCode = err.response.statusCode;
```
Comment (kabirbaidhya) [inline monitor/src/services/status.js]:
```ts
const { statusCode } = err.response;
```

**Pull request merged**